port/win/win_thread.h

```diff
@@ -0,0 +1,117 @@
+//  -- is that it dynamically allocates its internals that are automatically freed when
+//     the thread terminates and not on the destruction of the object. This makes
+//     it difficult to control the source of memory allocation
+//  - This implements Pimpl so we can easily replace the guts of the object in our
+//     private version if necessary.
+class WindowsThread {
+
+  struct Data;
+
+  std::unique_ptr<Data>  data_;
+  unsigned int           th_id_;
+
+  void Init(std::function<void()>&&);
+
+public:
+
+  typedef void* native_handle_type;
+
+  // Consstruct with no thread
+  WindowsThread();
+
+  // Template constructor
+  // 
+  // This templated constructor accomplishes several things
+  //
+  // - Allows the class as whole to be not a template
+  //
+  // - take "universal" references to support both lvalues and rvalues
+  //
+  // - because this constructor is a catchall case in many respects it may
+  //    prevent us from using both the default ctor, the move ctor. Also it may circumvent
+  //    copy ctor deletion. To work around this we make sure this one has at least one argument
+  //    and eliminate it from the overload  selection when WindowsThread is the first argument.
+  //
+  // - construct with Fx(Ax...) with a variable number of types/arguments.
+  //
+  // - Gathers together the callable object with its argments and constructs a single callable
+  //   entity
+  //
+  // - Makes use of std::function to convert it to a specificnon-template dependent type that both
+  //    checks the signature comformance to ensure that all of the necessary arguments are provided
+  //    and allows pimpl implementation.
+  template<class Fn,
+    class... Args,
+    class = typename std::enable_if<
+      !std::is_same<typename std::decay<Fn>::type,
+                    WindowsThread>::value>::type>
+  explicit WindowsThread(Fn&& fx, Args&&... ax) :
+      WindowsThread() {
+
+    // Use binder to create a single collable entity
+    auto binder = std::bind(std::forward<Fn>(fx),
+      std::forward<Args>(ax)...);
+    // Use std::function to take advantage of the type erasure
+    // so we can still hide implementation within pimpl
+    // This also makes sure that the binder signarture is compliant
+    std::function<void()> target = binder;
+
+    Init(std::move(target));
+  }
+
+
+  ~WindowsThread();
+
+  WindowsThread(const WindowsThread&) = delete;
+
+  WindowsThread& operator=(const WindowsThread&) = delete;
+
+  WindowsThread(WindowsThread&&) noexcept;
+
+  WindowsThread& operator=(WindowsThread&&) noexcept;
+
+  bool joinable() const;
+
+  unsigned int get_id() const { return th_id_; }
+
+  native_handle_type native_handle() const;
+
+  static unsigned hardware_concurrency();
+
+  void join();
+
+  bool detach();
+
+  void swap(WindowsThread&);
+};
+}
+}
+
+namespace std {
+  inline
+  void swap(rocksdb::port::WindowsThread& th1, rocksdb::port::WindowsThread& th2) {
+    th1.swap(th2);
+  }
+}
+
```

Comment:
Our format checker said lots of the lines are more than 80 bytes, and the ending bracket of namespace should followed by " // namespace <namespace>".

Since it is Windows port, it's your call to whether fix them.